src/main/java/appeng/parts/p2p/ItemP2PTunnelPart.java

```diff
@@ -118,36 +84,43 @@ public ItemStack insertItem(int slot, @Nonnull ItemStack stack, boolean simulate
                 int overflow = amountPerOutput == 0 ? amount : amount % amountPerOutput;
 
                 for (ItemP2PTunnelPart target : ItemP2PTunnelPart.this.getOutputs()) {
-                    final IItemHandler output = target.getAttachedItemHandler();
-                    final int toSend = amountPerOutput + overflow;
-                    final ItemStack fillWithItemStack = stack.copy();
-                    fillWithItemStack.setCount(toSend);
-
-                    ItemStack received = ItemStack.EMPTY;
+                    try (CapabilityGuard capabilityGuard = target.getAdjacentCapability()) {
+                        final IItemHandler output = capabilityGuard.get();
+                        final int toSend = amountPerOutput + overflow;
 
-                    for (int i = 0; i < output.getSlots(); i++) {
-                        received = output.insertItem(i, fillWithItemStack, simulate);
-
-                        if (received.isEmpty()) {
+                        if (toSend <= 0) {
+                            // Both overflow and amountPerOutput are 0, so they will be for further outputs as well.
                             break;
                         }
-                    }
 
-                    overflow = received.getCount();
-                    remainder.setCount(remainder.getCount() - toSend - received.getCount());
+                        // So the documentation says that copying the stack should not be necessary because it is not
+                        // supposed to be stored or modifed by insertItem. However, ItemStackHandler will gladly store
+                        // the stack so we need to do a defensive copy. Forgecord says this is the intended behavior,
+                        // and the documentation is wrong.
+                        ItemStack stackCopy = stack.copy();
+                        stackCopy.setCount(toSend);
+                        final int sent = toSend - ItemHandlerHelper.insertItem(output, stackCopy, simulate).getCount();
 
-                    if (remainder.isEmpty()) {
-                        break;
+                        overflow = toSend - sent;
+                        remainder -= sent;
                     }
                 }
 
                 if (!simulate) {
-                    ItemP2PTunnelPart.this.queueTunnelDrain(PowerUnits.RF, stack.getCount() - remainder.getCount());
+                    ItemP2PTunnelPart.this.queueTunnelDrain(PowerUnits.RF, amount - remainder);
                 }
             } catch (GridAccessException ignored) {
             }
 
-            return remainder;
+            if (remainder == stack.getCount()) {
+                return stack;
+            } else if (remainder == 0) {
+                return ItemStack.EMPTY;
```

Comment:
Cannot look it up currently, but I remember the helpers provided by forge usually returning `ItemStack.EMPTY` in case of it being empty and not the original stack set to a stacksize of 0. Not sure if this is defined somewhere as contract. This applies mostly in regards of the first condition, but currently this should never be 0. See the previous comment about a fast return.

Also similar to to capturing a negative `toSend`, what happens in case something returns negative amounts and the remainder is actually greater than `stack.getCount()`?

Reply:
Nobody checks these methods, so I don't think it makes sense for us to check them - the forge helper will fail silently, we might not even be able to detect that.